src/mscorlib/src/System/Guid.cs

```diff
@@ -365,214 +365,287 @@ public static bool TryParseExact(String input, String format, out Guid result)
+                    hasParenthesesInString = true;
                 }
             }
 
-            // Check for parenthesis
-            bool parenthesisExistInString = (guidString.IndexOf('(', 0) >= 0);
- 
-            if (parenthesisExistInString) {
-                if ((flags & (GuidStyles.AllowParenthesis | GuidStyles.RequireParenthesis)) == 0) {
-                    // parenthesis are not allowed
-                    result.SetFailure(ParseFailureKind.Format, "Format_GuidUnrecognized");
-                    return false;
-                }
+            if (!hasDashesInString && (flags & GuidStyles.RequireDashes) != 0)
+            {
+                // dashes are required
+                result.SetFailure(ParseFailureKind.Format, "Format_GuidUnrecognized");
+                return false;
             }
-            else {
-                if ((flags & GuidStyles.RequireParenthesis) != 0) {
-                    // parenthesis are required
-                    result.SetFailure(ParseFailureKind.Format, "Format_GuidUnrecognized");
-                    return false;
-                }
+            if (!hasBracesInString && (flags & GuidStyles.RequireBraces) != 0)
+            {
+                // braces are required
+                result.SetFailure(ParseFailureKind.Format, "Format_GuidUnrecognized");
+                return false;
+            }
+            if (!hasParenthesesInString && (flags & GuidStyles.RequireParenthesis) != 0)
+            {
+                // parenthesis are required
+                result.SetFailure(ParseFailureKind.Format, "Format_GuidUnrecognized");
+                return false;
             }
 
-            try {
+            try
+            {
                 // let's get on with the parsing
-                if (dashesExistInString) {
+                if (hasDashesInString)
+                {
                     // Check if it's of the form [{|(]dddddddd-dddd-dddd-dddd-dddddddddddd[}|)]
-                    return TryParseGuidWithDashes(guidString, ref result);
+                    return TryParseGuidWithDashes(g, start, length, ref result);
                 }
-                else if (bracesExistInString) {
+                else if (hasBracesInString)
+                {
                     // Check if it's of the form {0xdddddddd,0xdddd,0xdddd,{0xdd,0xdd,0xdd,0xdd,0xdd,0xdd,0xdd,0xdd}}
-                    return TryParseGuidWithHexPrefix(guidString, ref result);
+                    return TryParseGuidWithHexPrefix(g, start, length, ref result);
                 }
-                else {
+                else
+                {
                     // Check if it's of the form dddddddddddddddddddddddddddddddd
-                    return TryParseGuidWithNoStyle(guidString, ref result);
+                    return TryParseGuidWithNoStyle(g, start, length, ref result);
                 }
             }
-            catch(IndexOutOfRangeException ex) {
+            catch (IndexOutOfRangeException ex)
```

Comment:
@ellismg Is catching this IndexOutOfRangeException and the ArgumentException below necessary, or actually hiding a bug that has the potential to happen?